includes/data-port/class-sensei-import-csv-reader.php

```diff
@@ -0,0 +1,286 @@
+					$lines[] = array_combine( $columns, $indexed_line );
+				}
+			} else {
+				$lines[] = [];
+			}
+
+			if ( $this->file->eof() ) {
+				break;
+			}
+
+			$this->file->next();
+		}
+
+		if ( $this->file->eof() ) {
+			$this->is_completed = true;
+		}
+
+		$this->completed_lines += $lines_processed;
+
+		return $lines;
+	}
+
+	/**
+	 * Get the column names of the file.
+	 *
+	 * @return string[]
+	 */
+	private function get_column_names() {
+		$this->file->seek( 0 );
+
+		$column_names = $this->file->current();
+
+		if ( empty( $column_names ) ) {
+			return [];
+		}
+
+		// Make the column names of the CSV file case insensitive.
+		return array_map(
+			function ( $name ) {
+				return strtolower( trim( $name ) );
+			},
+			$column_names
+		);
+	}
+
+	/**
+	 * Whether the reading of the file is completed.
+	 *
+	 * @return bool
+	 */
+	public function is_completed() {
+		return $this->is_completed;
+	}
+
+	/**
+	 * The number of lines that have already been read.
+	 *
+	 * @return int
+	 */
+	public function get_completed_lines() {
+		return $this->completed_lines;
+	}
+
+	/**
+	 * The number of total lines in the file (the first line is not included).
+	 *
+	 * @return int
+	 */
+	public function get_total_lines() {
+		return $this->total_lines;
+	}
+
+	/**
+	 * Validate a CSV file.
+	 *
+	 * @param string $file_path        The file path.
+	 * @param array  $required_columns The columns that the CSV file is required to have.
+	 * @param array  $optional_columns The columns that are optional.
+	 *
+	 * @return bool|WP_Error
+	 */
+	public static function validate_csv_file( $file_path, $required_columns, $optional_columns ) {
+		if ( ! is_readable( $file_path ) ) {
+			return new WP_Error(
+				'sensei_data_port_job_unreadable_file',
+				__( 'Uploaded file could not be opened.', 'sensei-lms' )
+			);
+		}
+
+		try {
+			$reader = new Sensei_Import_CSV_Reader( $file_path );
+		} catch ( Exception $e ) {
+			return new WP_Error(
+				'sensei_data_port_job_unreadable_file',
+				$e->getMessage()
+			);
+		}
+
+		$columns = $reader->get_column_names();
+		if ( empty( $columns ) ) {
+			return new WP_Error(
+				'sensei_data_port_job_invalid_file',
+				__( 'Uploaded file was not a valid CSV.', 'sensei-lms' )
+			);
+		}
+
+		$has_required_columns     = array_intersect( $required_columns, $columns );
+		$missing_required_columns = array_diff( $required_columns, $has_required_columns );
+
+		if ( ! empty( $missing_required_columns ) ) {
+			return new WP_Error(
+				'sensei_data_port_job_missing_columns',
+				sprintf(
+					// translators: Placeholder is list of columns that are missing.
+					_n(
+						'Source file is missing the required column: %s',
+						'Source file is missing the required columns: %s',
+						count( $missing_required_columns ),
+						'sensei-lms'
+					),
+					implode( ', ', $missing_required_columns )
+				)
+			);
+		}
+
+		$unknown_columns = array_diff( $columns, $required_columns, $optional_columns );
```

Comment:
Should we actually care about unknown columns? I feel like it is fine. We could add it as a warning maybe in the job log.

Reply:
The senario that I am worried about is that a simple typo on an optional column name could possibly ruin the data for the whole import. 

On the other side there might be the case where a user did an export from another system, mapped the columns and left the ones that are unmapped in the CSV file. In that case it would be good if we ignored these. Not sure if this senario is possible though.

Reply:
In theory, wouldn't they be able to run it again with the corrected column name and the previously imported items would get fixed?

Reply:
I think for now I'm fine with this. Until we add the mapping step in v2, we're asking for a pretty rigid format.